common/app/model/ShareLinks.scala

```diff
@@ -21,15 +22,24 @@ final case class ShareLinks(
 ) {
   private def shareLink(shareType: String, elementId: Option[String] = None, mediaPath: Option[String] = None, shortLinkUrl: String, webLinkUrl: String, title: String): Option[ShareLink] = {
 
-    def shareCampaignUrl(campaign: String, elementId: Option[String]) = {
-      elementId.map { block => s"$shortLinkUrl/$campaign#$block" } getOrElse s"$shortLinkUrl/$campaign"
+    def createShareLinkUrl(campaign: Option[String], elementId: Option[String]): String = {
+      val campaignParam = campaign.flatMap(ShortCampaignCodes.getFullCampaign(_))
+      val queryParams: Map[String, String] = Map(
+        "page" -> elementId.map(id => s"with:$id"),
+        "CMP" -> campaignParam
+      )
+        .filter(_._2.isDefined)
+        .map { case (k, v) => (k, v.getOrElse("")) }
+
+      webLinkUrl.appendQueryParams(queryParams) + elementId.map(id => s"#${id}").getOrElse("")
     }
 
-    lazy val facebook = shareCampaignUrl("sfb", elementId).urlEncoded
-    lazy val googlePlus = shareCampaignUrl("sgp", elementId).urlEncoded
-    lazy val link = shareCampaignUrl("sbl", elementId).urlEncoded
-    lazy val twitter = shareCampaignUrl("stw", elementId).urlEncoded
-    lazy val whatsapp = shareCampaignUrl("swa", elementId)
+    lazy val facebook = createShareLinkUrl(Some("sfb"), elementId).urlEncoded
+    lazy val googlePlus = createShareLinkUrl(Some("sgp"), elementId).urlEncoded
+    lazy val link = createShareLinkUrl(Some("sbl"), elementId).urlEncoded
+    lazy val twitter = createShareLinkUrl(Some("stw"), elementId).urlEncoded
+    lazy val whatsapp = createShareLinkUrl(Some("swa"), elementId)
+    lazy val pinterest = createShareLinkUrl(None, elementId).urlEncoded
```

Comment:
And linkedIn. Will follow up

On Wed, 3 Feb 2016 at 11:06 Richard Nguyen <notifications@github.com> wrote:

> In common/app/model/ShareLinks.scala
> <https://github.com/guardian/frontend/pull/11740#discussion_r51707151>:
>
> >      }
> >
> > -    lazy val facebook = shareCampaignUrl("sfb", elementId).urlEncoded
> > -    lazy val googlePlus = shareCampaignUrl("sgp", elementId).urlEncoded
> > -    lazy val link = shareCampaignUrl("sbl", elementId).urlEncoded
> > -    lazy val twitter = shareCampaignUrl("stw", elementId).urlEncoded
> > -    lazy val whatsapp = shareCampaignUrl("swa", elementId)
> > +    lazy val facebook = createShareLinkUrl(Some("sfb"), elementId).urlEncoded
> > +    lazy val googlePlus = createShareLinkUrl(Some("sgp"), elementId).urlEncoded
> > +    lazy val link = createShareLinkUrl(Some("sbl"), elementId).urlEncoded
> > +    lazy val twitter = createShareLinkUrl(Some("stw"), elementId).urlEncoded
> > +    lazy val whatsapp = createShareLinkUrl(Some("swa"), elementId)
> > +    lazy val pinterest = createShareLinkUrl(None, elementId).urlEncoded
>
> I think this needs a campaign code too.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/frontend/pull/11740/files#r51707151>.
>

-- 

------------------------------
This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396